### PR TITLE
SUPPORT-84787 add client walles permissions management

### PR DESCRIPTION
**Pull request by panchenkodv**
Update needed according to ticket SUPPORT-84787

### CODE REVIEW

File: src/Paysera/WalletApi/Client/WalletClient.php
Comment (pelanis) [inline src/Paysera/WalletApi/Client/WalletClient.php:1027]: why data are not mapped here?

There should be some permissions Entity. Permissions might have also more information, then only the scope itself and I doubt it's also important permissions type.


We have  Paysera_WalletApi_Entity_ClientPermissions maybe we could do something similar, maybe better, but it should be not just mixed|null returned

File: src/Paysera/WalletApi/Entity/Client.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/Client.php:286]: if there is a variable with set permissions, then we should also have the option to add by one entry.

Another question what does it mean walletsPermissions? it's a bit unclear what it sets.

It looked like Client Permissions to the specific wallet, so maybe could be like: ClientPermissionsToWallets, ClientPermissionsByWallets... what do you think?

File: src/Paysera/WalletApi/Entity/Client.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/Client.php:284]: return self

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/ClientWalletPermissions.php:12]: check Paysera_WalletApi_OAuth_Consumer

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/ClientWalletPermissions.php:27]: = [];

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/ClientWalletPermissions.php:58]: it returns self

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/ClientWalletPermissions.php:41]: it doesn't look like Paysera_WalletApi_Entity_ClientWalletPermissions can be without walletId, so it must be in the constructor, otherwise, it can be, that we would have permissions, but no id set...

by the way, if we would be just reusing what we have, maybe we could survive without constructor, as it's already like that...
Paysera_WalletApi_Entity_Wallet

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2425]: why there is such case? can't it be other permissions?

In case we say, that it's just for balance and statements, it must be clear. In case it's some "security" measure, then it's wrong, as everyone can change such "case"

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2083]: review places, where you use array(), it must be []

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2443]: $walletPermissions->getWalletId() === null, if walletId will be 0, it's still id, that's the reason why we should always use strict comparisons

**Review by pelanis — CHANGES_REQUESTED** [review]

File: src/Paysera/WalletApi/Client/WalletClient.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Client/WalletClient.php:1027]: There is mixed data of available wallets and existing permissions and looks like
[{"id": 1, "account": "EVP1","balance": false,"statement": false}, {"id": 2, "account": "EVP2","balance": true,"statement": false}, {"id": 3, "account": "EVP3","balance": true,"statement": true}]

File: src/Paysera/WalletApi/Mapper.php
Comment (ddimoss) [inline src/Paysera/WalletApi/Mapper.php:2039]: https://github.com/paysera/php-style-guide#checking-things-explicitly

we use count($array) > 0 to check if array is not empty and not !empty($array), as we do not want to check whether $array is 0, false, '' or even not defined at all (in which case IDE would possibly hide some warnings that could help noticing possible bugs).

Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php:2039]: In my case it can be undefined, so I need use if (isset($data['permissions_to_wallets']) && count($data['permissions_to_wallets']) > 0) ?

File: src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php:172]: additional explanation Grant gives no info, same as other Revoke or etc, we avoid adding comments into our code, in case you are going to add it, this means something is unclear and we should review the code, as it requires explanations.

it's already clear enough, that revoke is Revoke or isGrant checking if Granted

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2040]: actually it would be more correct to say, that it's $clientPermissionToWallet

File: src/Paysera/WalletApi/Entity/Client.php
Comment (pelanis) [inline src/Paysera/WalletApi/Entity/Client.php:61]: we should set it in the constructor, not where we name variable.

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2449]: you have a check with id, would it be fine if accountNumber or scopes not set?

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:16]: We have a test for Mapper, as we added new methods into the Client, I think would be also good to add mapper test, so we would be sure it's returning the expected results not random data.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:23]: is it intended to launc parent::setUp(); at the end? in some cases it could overwrite the values you set.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:159]: let's use not snake cased format, just write in words something like... 'No scopes'

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:205]: what would happen in case I use the unknown scope? would it allow to use it?

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:186]: if you manually enter values in each case, would be good to use it and change values in there.

**Review by pelanis — CHANGES_REQUESTED** [review]

File: src/Paysera/WalletApi/Client/WalletClient.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Client/WalletClient.php]: I've solved it and added mapper.

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2451]: wouldn't it be better to return Wallet structure of data?

Also, make sure, that on the wallet-api side the scope permissions are correctly assigned, because we show not only permissions, but also account info (account number is account info).

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2409]: it would be better, if besides of using generic variables like array/list/string/data you would be using variable, so in the specific case it should be $clientPermissionsToWallet

File: tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php:5]: We don't write tests to Entitites, we cover entity with the tests in services, so it's ok to cover with the tests Mapper, Manager and similar things.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [reply tests/Paysera/WalletApi/MapperTest.php:16]: how about this? is it done?

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [reply tests/Paysera/WalletApi/MapperTest.php:24]: as I understand it's resolved.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [inline tests/Paysera/WalletApi/MapperTest.php:235]: I see in each test 'owner' is null, do we even need that value? if yes, maybe would be also good to test if it's not null.

**Review by pelanis — CHANGES_REQUESTED** [review]

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php:235]: In my case I no need this value, but here https://github.com/paysera/lib-wallet-php-client/blob/a574f2eb95374b4ddae1d8d9a03b694558dda3c4/src/Paysera/WalletApi/Mapper.php#L1206 I'll get an error if not provide this key.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php:24]: Yes, I moved it to the top.

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [reply tests/Paysera/WalletApi/MapperTest.php:235]: this means it must be something like owner1 if it's string or 1 if int, by the way, that's the point, that I think we are missing owner.

Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php:235]: Actually we haven't owner info in this point and it's unnecessary information. But if I provide some id like 1 that means that wallet belongs to client with id 1. It could cause unpredictable behavior and bugs.

File: src/Paysera/WalletApi/Entity/Client.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/Client.php]: Done

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientWalletPermissions.php]: Done

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientWalletPermissions.php]: Done

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientWalletPermissions.php]: Done

File: src/Paysera/WalletApi/Entity/ClientWalletPermissions.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientWalletPermissions.php]: Done

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Done

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Done

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Done

File: src/Paysera/WalletApi/Entity/Client.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/Client.php]: Done

File: src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php]: Done

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php]: Done

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php]: Done

File: tests/Paysera/WalletApi/MapperTest.php
Comment (pelanis) [reply tests/Paysera/WalletApi/MapperTest.php:235]: In real situation there can't be wallet without owner

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Done

File: tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php:5]: May I leave this test? It's too pity to delete it :(

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php:2459]: In this case this structure is better than just wallet structure.

File: src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/ClientPermissionsToWallet.php]: Done

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php:24]: Yes, I added tests to mapper new tests for encoding and decoding client

File: tests/Paysera/WalletApi/MapperTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/MapperTest.php:853]: In my case I need only wallet ID as I resolved wallet by ID on wallet API side.

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php:2459]: Also I've added checking if user has permissions to read wallets list in the first step in method on wallet-api side

**Review by pelanis — CHANGES_REQUESTED** [review]

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: I've changed naming from `id` to `wallet_id` and from `account` to `account_number`. I have no need to encode whole wallet structure to array and send it to wallet API as I resolve wallet info by wallet ID on wallet API side.

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Removed name duplication

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php:2406]: Done

File: tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/Entity/ClientPermissionsToWalletTest.php]: Done

**Comment by panchenkodv** [comment]
I've resolved all comments

File: src/Paysera/WalletApi/Entity/Client.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Entity/Client.php]: Done

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Added accountNumber checking

**Review by pelanis — CHANGES_REQUESTED** [review]

File: tests/Paysera/WalletApi/Entity/ClientTest.php
Comment (panchenkodv) [reply tests/Paysera/WalletApi/Entity/ClientTest.php]: I moved it to Mapper, so it's duplicating. I removed it.

File: composer.json
Comment (panchenkodv) [reply composer.json]: It was need for WalletTrait. But as I removed entity tests trait does not need anymore and this changes as well. I reverted this changes.

File: src/Paysera/WalletApi/Mapper.php
Comment (pelanis) [inline src/Paysera/WalletApi/Mapper.php:2432]: we should use **new Paysera_WalletApi_Entity_ClientPermissionsToWallet()**

File: src/Paysera/WalletApi/Client/WalletClient.php
Comment (pelanis) [inline src/Paysera/WalletApi/Client/WalletClient.php:1028]: is it ok, that there such endpoint doesn't exist?

**Review by pelanis — CHANGES_REQUESTED** [review]

File: src/Paysera/WalletApi/Client/WalletClient.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Client/WalletClient.php:1028]: It exists in branch SUPPORT-84787_add_oauth_permission_manager in wallet-api project. We will deploy it before merging this PR.

File: src/Paysera/WalletApi/Mapper.php
Comment (panchenkodv) [reply src/Paysera/WalletApi/Mapper.php]: Done.

**Pull request merged**
